dr_ensenso_node/src/ensenso.cpp

```diff
@@ -483,6 +487,64 @@ class EnsensoNode: public ros::NodeHandle {
 		return true;
 	}
 
+	bool onGetCameraParams(dr_ensenso_msgs::GetCameraParams::Request & req, dr_ensenso_msgs::GetCameraParams::Response & res) {
+		if (true
+			&& (req.camera) == dr_ensenso_msgs::GetCameraParams::Request::MONO
+			&& (req.camera) == dr_ensenso_msgs::GetCameraParams::Request::LEFT
+			&& (req.camera) == dr_ensenso_msgs::GetCameraParams::Request::RIGHT
+		) {
+			ROS_ERROR_STREAM("Invalid camera type " << req.camera << " requested.");
+			return false;
+		}
+
+		if (req.camera == dr_ensenso_msgs::GetCameraParams::Request::MONO && !ensenso_camera->hasMonocular()) {
+			ROS_ERROR_STREAM("No monocular camera available!");
+			return false;
+		}
+
+		cv::Mat mat;
+		switch (req.type) {
+			case dr_ensenso_msgs::GetCameraParams::Request::CAMERA_MATRIX: {
+				NxLibItem item = (req.camera == dr_ensenso_msgs::GetCameraParams::Request::MONO)
+				? (*(ensenso_camera->nativeMonocular()))[itmCalibration]
+				: ensenso_camera->native()[itmCalibration][itmMonocular][req.camera == "Left" ? itmLeft : itmRight];
+				mat = toCameraMatrix(item);
+				break;
+			} case dr_ensenso_msgs::GetCameraParams::Request::DISTORTION_MATRIX: {
+				NxLibItem item = (req.camera == dr_ensenso_msgs::GetCameraParams::Request::MONO)
+				? (*(ensenso_camera->nativeMonocular()))[itmCalibration]
+				: ensenso_camera->native()[itmCalibration][itmMonocular][req.camera == "Left" ? itmLeft : itmRight];
+				mat = toDistortionParameters(item);
+				break;
+			} case dr_ensenso_msgs::GetCameraParams::Request::RECTIFICATION_MATRIX: {
+				if (req.camera == dr_ensenso_msgs::GetCameraParams::Request::MONO) {
+					ROS_ERROR_STREAM("Request for mono camera rectification matrix is unsupported!");
+					return false;
+				}
+				mat = toRectificationMatrix(ensenso_camera->native()[itmCalibration][itmStereo][req.camera == "Left" ? itmLeft : itmRight]);
+				break;
+			} case dr_ensenso_msgs::GetCameraParams::Request::PROJECTION_MATRIX: {
+				if (req.camera == dr_ensenso_msgs::GetCameraParams::Request::MONO) {
+					ROS_ERROR_STREAM("Request for mono camera projection matrix is unsupported!");
+					return false;
+				}
+				mat = toProjectionMatrix(ensenso_camera->native()[itmCalibration], req.camera);
+				break;
+			} default: {
+				ROS_ERROR_STREAM("Invalid type requested!");
+				return false;
+			}
+		}
+
+		for (int i = 0; i < mat.rows; i++) {
+			for (int j = 0; j < mat.cols; j++) {
```

Comment:
This is now a column major matrix. `sensor_msgs::CameraInfo` uses row major storage. I think it's best not to deviate from that.

Reply:
I go from left to right and top to bottom, isn't that actually row major?

https://en.wikipedia.org/wiki/Row-_and_column-major_order

Reply:
ah, sorry, I mixed up `int j = 0; j < mat.cols; j++` with looping through a column rather than looping over the columns. You're right, it is row major already.